database.go

```diff
@@ -737,3 +740,203 @@ func marshalTags(tags map[string]string) []byte {
+type dbi struct {
+	server *Server
+	db     *database
+}
+
+// MatchSeries returns a list of series data ids matching a name and tags.
+func (dbi *dbi) MatchSeries(name string, tags map[string]string) (a []uint32) {
+	// Find measurement by name.
+	m := dbi.db.measurements[name]
+	if m == nil {
+		return nil
+	}
+
+	// Match each series on the measurement by tagset.
+	// TODO: Use paul's fancy index.
+	for _, s := range m.seriesByID {
+		if s.match(tags) {
+			a = append(a, s.ID)
+		}
+	}
+	return
+}
+
+// SeriesTagValues returns a slice of tag values for a series.
+func (dbi *dbi) SeriesTagValues(seriesID uint32, keys []string) []string {
+	// Find series by id.
+	s := dbi.db.series[seriesID]
+
+	// Lookup value for each key.
+	values := make([]string, len(keys))
+	for i, keys := range keys {
+		values[i] = s.Tags[keys]
+	}
+	return values
+}
+
+// Field returns the id and data type for a series field.
+// Returns id of zero if not a field.
+func (dbi *dbi) Field(name, field string) (fieldID uint8, typ influxql.DataType) {
+	// Find measurement by name.
+	m := dbi.db.measurements[name]
+	if m == nil {
+		return 0, influxql.Unknown
+	}
+
+	// Find field by name.
+	f := m.FieldByName(field)
+	if f == nil {
+		return 0, influxql.Unknown
+	}
+
+	return f.ID, f.Type
+}
+
+// CreateIterator returns an iterator given a series data id, field id, & field data type.
+func (dbi *dbi) CreateIterator(seriesID uint32, fieldID uint8, typ influxql.DataType, min, max time.Time, interval time.Duration) influxql.Iterator {
+	// TODO: Add retention policy to the arguments.
+
+	// Create an iterator to hold the transaction and series ids.
+	itr := &iterator{
+		seriesID: seriesID,
+		fieldID:  fieldID,
+		typ:      typ,
+		imin:     -1,
+		interval: int64(interval),
+	}
+	if !min.IsZero() {
+		itr.min = min.UnixNano()
+	}
+	if !max.IsZero() {
+		itr.max = max.UnixNano()
+	}
+
+	// Retrieve the policy.
+	// Ignore if there are no groups created on the retention policy.
+	rp := dbi.db.policies[dbi.db.defaultRetentionPolicy]
+	if len(rp.groups) == 0 {
+		return itr
+	}
+
+	// Find all shards which match the the time range and series id.
+	// TODO: Support multiple groups.
+	g := rp.groups[0]
+
+	// Ignore shard groups that our time range does not cross.
+	if !timeBetweenInclusive(g.StartTime, min, max) &&
+		!timeBetweenInclusive(g.EndTime, min, max) {
+		return itr
+	}
+
+	// Find appropriate shard by series id.
+	sh := g.ShardBySeriesID(seriesID)
+
+	// Open a transaction on the shard.
+	tx, err := sh.store.Begin(false)
+	assert(err == nil, "read-only tx error: %s", err)
+	itr.tx = tx
+
+	// Open and position cursor.
+	b := tx.Bucket(u32tob(seriesID))
+	if b != nil {
+		cur := b.Cursor()
+		itr.k, itr.v = cur.Seek(u64tob(uint64(itr.min)))
+		itr.cur = cur
+	}
+
+	return itr
+}
+
+// iterator represents a series data iterator for a shard.
+// It can iterate over all data for a given time range for multiple series in a shard.
```

Comment:
It's single series right now. I need to refactor the engine to group series together and then I can fix the `influxdb.iterator`.